### DISCUSSION TITLE
Action API

### DISCUSSION THREAD

**Discussion by pilcrowonpaper** [withastro/roadmap#427] [Proposal]
Add a built in way to handle form submissions (actions) and implement forms with progressive enhancements

This was originally based on [my implementation](https://github.com/pilcrowOnPaper/astro-forms) but it's now updated to be based on an actual library: https://github.com/pilcrowOnPaper/astro-form-actions

## Why?

While Astro is a static-first framework, I truly believe it can be great framework for building dynamic websites, including those with _some_ user interactions. Something like a forum with authentication. It's mostly rendering dynamic content with a few forms for login, making posts, and replying to posts. SSR is already implemented and it already has an API for handling requests/responses in pages, as well as support for API routes. The only major thing missing now is a way to handle forms. Frameworks like [SvelteKit](https://kit.svelte.dev/docs/form-actions) and [Remix](https://remix.run/docs/en/v1/route/action) makes it easy to with their own `action` API, and I think Astro could implement one as well.

One may forget to implement proper CSRF protection as well. 

### Handling form submissions

Pages can export an asynchronous `action` function, which will handle form submissions, specifically `POST` requests with a `Content-Type` header of `multipart/form-data` or `application/x-www-form-urlencoded`. This `action()` will run before any server side code in the page file.

It only accept a single argument: `formData`. This will be the `FormData` object sent with the request, which includes both string and file input. The request, response, url, etc can be accessed from the Astro global object.

`action()` must return one of 3 responses: `Resolved` (success), `Rejected` (error), `Redirected` (redirect). These are created using `resolve()`, `reject()`, and `redirect()` respectively. `resolve()` takes a POJO, `reject()` takes a status code and POJO, and `redirect()` takes a status code and location. All three functions can be imported from `astro/actions` (some module).

```ts
import { resolve, reject, redirect } from "astro/actions";

export const action = async (formData: FormData) => {
  if (isInvalid)
    return reject(400, {
      message: "nope",
    });
  if (redirectResponse) return redirect(302, "/");
  return resolve({
    message: "success!",
  });
};
```

After handling the form submission, if the request was sent by the browser (native way), it will immediately return a redirect response on redirect or simply re-render the page on success or error. If the request was sent using client-side JavaScript (`fetch()`), it will return a JSON response immediately regardless of the response type. The only time it will modify the status code of `Astro.response` is if it's an error response for a native form submission. 

The result of action (`resolve()`) can be retrieved using `Astro.useActionResult<Action>()`. 

```ts
const result = Astro.useActionResult<typeof action>();
```

This will take the type of `action` as the generic, and return the following type:

```ts
type ActionResult = {
  type: "resolved" | "rejected";
  body: null | Body; // resolved data
  error: null | ErrorData; // error data from 2nd params of reject()
  inputValues: Record<string, string | undefined>; // object from formData, without file inputs
} | null;
```

Type `Body `will be the argument for `resolved()`, and `ErrorData` will be the 2nd argument for `rejected()`. See library for implementation.

### Submitting forms

Forms can be submitted as is (handled by the browser), or using the `submitForm()` function imported from `astro/forms`. In either case, a form element is used.

`submitForm()` is a generic function that can be used with any JS framework. It takes 2 arguments: the html `form` element, and a optional redirect handler. Since the redirect will be handled if JavaScript, it may be useful to use client side navigation rather than server side navigation. The default behavior is to use `window.location.href` to redirect.

```ts
const submitForm: <Data extends {}>(
  formElement: FormElement,
  redirectHandler?: () => void
) => Promise<ActionResult<Data>>;
```

Return type is as follow (`ActionResult` without `inputValues`): 

```ts
type Result = {
  type: "resolved" | "rejected" | "redirect";
  body: null | Body; // resolved data
  error: null | ErrorData; // error data from 2nd params of reject()
}
```

## Constraints

- This API will only with `POST` forms. I don't think an API for `GET` is necessary though.

### Form component with progressive enhancements

This example uses Solid.

This will update the error message when it returns one, and will work regardless of the availability of client side JavaScript.

```tsx
import { submitForm } from "astro/forms";

export default (props: {
  errorMessage?: string; // error.message from actionResult
  inputValues?: Record<string, any>; // inputValues from actionResult
}) => {
  const [errorMessage, setErrorMessage] = createSignal(
    props.errorMessage ?? null
  );
  const handleSubmit = async (e) => {
    e.preventDefault();
    const { error } = await submitForm(e.currentTarget);
    if (error) setErrorMessage(error.message);
  };
  return (
    <form class="w-full" onSubmit={handleSubmit} action="/" method="post">
      <label for="message">Message</label>
      <input
        name="message"
        value={props.inputValues?.message ?? ""}
        id="message"
      />
      <input type="submit" />
      <Show when={errorMessage()}>
        <p>{errorMessage()}</p>
      </Show>
    </form>
  );
};
```

### Encoding type

The request should be cloned before doing anything with it.

Since the `formData()` method of native `Request` object only supports `application/x-www-form-urlencoded`, a parse must be used (eg. [`parse-multipart-data`](https://www.npmjs.com/package/parse-multipart-data) used by SolidStart). This is the only external dependency necessary.

See [here](https://github.com/pilcrowOnPaper/astro-forms/blob/842c01b76c710eab5a307548f7d49d872febb5dc/src/actions/server.ts#L33) for transforming form request body to `FormData`.

### Redirect responses

When using `fetch()` and a redirect response with a location header is returned, it will try to call a cross-site request to the redirect location. As such, responses for requests made with `fetch()` must not include the location header and the redirect location should be included in the response body.

The response body could look something like this:

```ts
type JSONResult<Data extends {}> = {
  type: "resolved" | "rejected" | "redirect";
  data: Data | null;
  error: Record<string, any> | null;
  redirect_location: string | null;
};
```

### CSRF protection

By default, actions should only accept requests coming from the same origin it's hosted. This is a **MUST** if this feature is to be implemented. This can be achieved using the `Origin` header, which all modern browsers support. It should be possible to disable feature, either in a config file or by passing on an option to `action()`, for those who want to support older browsers. 

```ts
if (request.headers.get('origin') !== url.origin) throw new Error();
```

### Type safety

See https://github.com/pilcrowOnPaper/astro-form-actions for how the result can be properly typed.

Types for the error should be declared in the form component, and TS will properly check the types when the `error` from the server is passed to it as props:
https://github.com/pilcrowOnPaper/astro-form-actions/wiki/3.-Working-with-types

**Comment by IshanKBG** [comment]
This seems interesting

**Comment by fprl** [comment]
I think this RFC + Endpoints v2 (to handle things like auth) are the missing piece of Astro SSR to adopt it before other frameworks.

**Reply by pilcrowonpaper** [reply]
What's "Endpoints v2"?

**Reply by fprl** [reply]
You can check [Exploration Endpoints v2](https://discord.com/channels/830184174198718474/1046817417365569577/1067887444441178113). I think it's the same as your [Astro middleware](https://github.com/withastro/roadmap/discussions/469) middleware.

**Comment by mfrachet** [comment]
Funny enough, coming from Remix, I was experimenting with something similar and was going to open this discussion 😊 . 

As of now, Astro is very good for initial load and "JavaScript sparkling" with the island architecture. However, the user's adventures do not stop after the page has finished loading.

This RFC looks very promising to me in the sense that it can be done in a framework-agnostic way and covers the subsequent navigation experience of a user. 

Kudos for the RFC, and I would love to see something similar landing in the tool 👏🏻 👏🏻

**Comment by lilnasy** [comment]
Love it!

**Comment by bholmesdev** [comment]
This is an awesome discussion! I'm happy to share that actions have landed as a stage 2 proposal ready to enter stage 3 soon. There are some key ideas from this thread that really resonate:
- An `Astro.useActionResult()` utility to get a type-safe result in your Astro frontmatter.
- First-class support for the `FormData` object.
- Progressive enhancement by POSTing to your current route.

I think all of these points should inform the final API. Closing to track here: https://github.com/withastro/roadmap/issues/898
